Calling `Node.store` while iterating over result of `QueryBuilder.iterall` or `QueryBuilder.iterdict` will throw an exception

For example, the following test:
```python
@pytest.mark.usefixtures('aiida_profile_clean')
def test_iterall_with_store(self):
    """Test that nodes can be mutated while being iterated using ``QueryBuilder.iterall``."""
    count = 10
    pks = []
    clone_pks = []

    for _ in range(count):
        node = orm.Data().store()
        pks.append(node.pk)

    # Ensure that batch size is smaller than the total rows yielded
    for [node] in orm.QueryBuilder().append(orm.Data).iterall(batch_size=2):
        clone = copy.deepcopy(node)
        clone.store()
        clone_pks.append(clone.pk)

    for pk, pk_clone in zip(pks, clone_pks):
        assert orm.load_node(pk) == orm.load_node(clone_pk)
```
excepts with the following exception:
```
tests/orm/test_querybuilder.py:1504: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
aiida/orm/nodes/node.py:466: in store
    self._store(with_transaction=with_transaction, clean=True)
aiida/orm/nodes/node.py:486: in _store
    self.base.caching.rehash()
aiida/orm/nodes/caching.py:74: in rehash
    self._node.base.extras.set(self._HASH_EXTRA_KEY, self.get_hash())
aiida/orm/nodes/caching.py:31: in get_hash
    if not self._node.is_stored:
aiida/orm/entities.py:263: in is_stored
    return self._backend_entity.is_stored
aiida/storage/psql_dos/orm/entities.py:83: in is_stored
    return self.model.id is not None
aiida/storage/psql_dos/orm/utils.py:84: in __getattr__
    if self.is_saved() and self._is_mutable_model_field(item) and not self._in_transaction():
aiida/storage/psql_dos/orm/utils.py:110: in is_saved
    return self._model.id is not None
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py:482: in __get__
    return self.impl.get(state, dict_)
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py:942: in get
    value = self._fire_loader_callables(state, key, passive)
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py:973: in _fire_loader_callables
    return state._load_expired(state, passive)
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/state.py:712: in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/loading.py:1451: in load_scalar_attributes
    result = load_on_ident(
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/loading.py:407: in load_on_ident
    return load_on_pk_identity(
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/loading.py:530: in load_on_pk_identity
    session.execute(
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/session.py:1713: in execute
    conn = self._connection_for_bind(bind)
/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/sqlalchemy/orm/session.py:1549: in _connection_for_bind
    TransactionalContext._trans_ctx_check(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'sqlalchemy.engine.util.TransactionalContext'>, subject = <sqlalchemy.orm.session.Session object at 0x7f1429bc7ac0>

    @classmethod
    def _trans_ctx_check(cls, subject):
        trans_context = subject._trans_context_manager
        if trans_context:
            if not trans_context._transaction_is_active():
>               raise exc.InvalidRequestError(
                    "Can't operate on closed transaction inside context "
                    "manager.  Please complete the context manager "
                    "before emitting further commands."
                )
E               sqlalchemy.exc.InvalidRequestError: Can't operate on closed transaction inside context manager.  Please complete the context manager before emitting further commands.
```

The commit 98fd0671814b6fc1381ec9e5b9e3d2165402d5f5 is probably related as that addressed a similar exception when mutating existing nodes. Probably this introduced this regression bug when storing new nodes though.

--- a/aiida/orm/implementation/nodes.py
+++ b/aiida/orm/implementation/nodes.py
@@ -184,13 +184,11 @@ class BackendNode(BackendEntity, BackendEntityExtrasMixin, metaclass=abc.ABCMeta
     def store(  # pylint: disable=arguments-differ
         self: BackendNodeType,
         links: Optional[Sequence['LinkTriple']] = None,
-        with_transaction: bool = True,
         clean: bool = True
     ) -> BackendNodeType:
         """Store the node in the database.
 
         :param links: optional links to add before storing
-        :param with_transaction: if False, do not use a transaction because the caller will already have opened one.
         :param clean: boolean, if True, will clean the attributes and extras before attempting to store
         """
 

--- a/aiida/orm/nodes/node.py
+++ b/aiida/orm/nodes/node.py
@@ -424,12 +424,10 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
         """
         return self.backend_entity.mtime
 
-    def store_all(self, with_transaction: bool = True) -> 'Node':
+    def store_all(self) -> 'Node':
         """Store the node, together with all input links.
 
         Unstored nodes from cached incoming linkswill also be stored.
-
-        :parameter with_transaction: if False, do not use a transaction because the caller will already have opened one.
         """
         if self.is_stored:
             raise exceptions.ModificationNotAllowed(f'Node<{self.pk}> is already stored')
@@ -440,11 +438,11 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
 
         for link_triple in self.base.links.incoming_cache:
             if not link_triple.node.is_stored:
-                link_triple.node.store(with_transaction=with_transaction)
+                link_triple.node.store()
 
-        return self.store(with_transaction)
+        return self.store()
 
-    def store(self, with_transaction: bool = True) -> 'Node':  # pylint: disable=arguments-differ
+    def store(self) -> 'Node':  # pylint: disable=arguments-differ
         """Store the node in the database while saving its attributes and repository directory.
 
         After being called attributes cannot be changed anymore! Instead, extras can be changed only AFTER calling
@@ -452,8 +450,6 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
 
         :note: After successful storage, those links that are in the cache, and for which also the parent node is
             already stored, will be automatically stored. The others will remain unstored.
-
-        :parameter with_transaction: if False, do not use a transaction because the caller will already have opened one.
         """
         from aiida.manage.caching import get_use_cache
 
@@ -477,9 +473,9 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
             same_node = self.base.caching._get_same_node() if use_cache else None  # pylint: disable=protected-access
 
             if same_node is not None:
-                self._store_from_cache(same_node, with_transaction=with_transaction)
+                self._store_from_cache(same_node)
             else:
-                self._store(with_transaction=with_transaction, clean=True)
+                self._store(clean=True)
 
             if self.backend.autogroup.is_to_be_grouped(self):
                 group = self.backend.autogroup.get_or_create_group()
@@ -487,16 +483,15 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
 
         return self
 
-    def _store(self, with_transaction: bool = True, clean: bool = True) -> 'Node':
+    def _store(self, clean: bool = True) -> 'Node':
         """Store the node in the database while saving its attributes and repository directory.
 
-        :param with_transaction: if False, do not use a transaction because the caller will already have opened one.
         :param clean: boolean, if True, will clean the attributes and extras before attempting to store
         """
         self.base.repository._store()  # pylint: disable=protected-access
 
         links = self.base.links.incoming_cache
-        self._backend_entity.store(links, with_transaction=with_transaction, clean=clean)
+        self._backend_entity.store(links, clean=clean)
 
         self.base.links.incoming_cache = []
         self.base.caching.rehash()
@@ -514,7 +509,7 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
                     f'Cannot store because source node of link triple {link_triple} is not stored'
                 )
 
-    def _store_from_cache(self, cache_node: 'Node', with_transaction: bool) -> None:
+    def _store_from_cache(self, cache_node: 'Node') -> None:
         """Store this node from an existing cache node.
 
         .. note::
@@ -542,7 +537,7 @@ class Node(Entity['BackendNode', NodeCollection], metaclass=AbstractNodeMeta):
             if key != Sealable.SEALED_KEY:
                 self.base.attributes.set(key, value)
 
-        self._store(with_transaction=with_transaction, clean=False)
+        self._store(clean=False)
         self._add_outputs_from_cache(cache_node)
         self.base.extras.set('_aiida_cached_from', cache_node.uuid)
 

--- a/aiida/storage/psql_dos/orm/groups.py
+++ b/aiida/storage/psql_dos/orm/groups.py
@@ -68,8 +68,7 @@ class SqlaGroup(entities.SqlaModelEntity[DbGroup], ExtrasMixin, BackendGroup):  
             try:
                 self.model.save()
             except Exception:
-                raise UniquenessError(f'a group of the same type with the label {label} already exists') \
-                    from Exception
+                raise UniquenessError(f'a group of the same type with the label {label} already exists') from Exception
 
     @property
     def description(self):
@@ -103,12 +102,6 @@ class SqlaGroup(entities.SqlaModelEntity[DbGroup], ExtrasMixin, BackendGroup):  
     @property
     def uuid(self):
         return str(self.model.uuid)
-
-    def __int__(self):
-        if not self.is_stored:
-            return None
-
-        return self._dbnode.id  # pylint: disable=no-member
 
     @property
     def is_stored(self):
@@ -220,7 +213,8 @@ class SqlaGroup(entities.SqlaModelEntity[DbGroup], ExtrasMixin, BackendGroup):  
                 session.execute(ins.on_conflict_do_nothing(index_elements=['dbnode_id', 'dbgroup_id']))
 
             # Commit everything as up till now we've just flushed
-            session.commit()
+            if not session.in_nested_transaction():
+                session.commit()
 
     def remove_nodes(self, nodes, **kwargs):
         """Remove a node or a set of nodes from the group.
@@ -268,7 +262,8 @@ class SqlaGroup(entities.SqlaModelEntity[DbGroup], ExtrasMixin, BackendGroup):  
                     statement = table.delete().where(clause)
                     session.execute(statement)
 
-            session.commit()
+            if not session.in_nested_transaction():
+                session.commit()
 
 
 class SqlaGroupCollection(BackendGroupCollection):

--- a/aiida/storage/psql_dos/orm/nodes.py
+++ b/aiida/storage/psql_dos/orm/nodes.py
@@ -211,7 +211,7 @@ class SqlaNode(entities.SqlaModelEntity[models.DbNode], ExtrasMixin, BackendNode
         self.model.attributes = clean_value(self.model.attributes)
         self.model.extras = clean_value(self.model.extras)
 
-    def store(self, links=None, with_transaction=True, clean=True):  # pylint: disable=arguments-differ
+    def store(self, links=None, clean=True):  # pylint: disable=arguments-differ
         session = self.backend.get_session()
 
         if clean:
@@ -223,12 +223,15 @@ class SqlaNode(entities.SqlaModelEntity[models.DbNode], ExtrasMixin, BackendNode
             for link_triple in links:
                 self._add_link(*link_triple)
 
-        if with_transaction:
+        if not session.in_nested_transaction():
             try:
                 session.commit()
             except SQLAlchemyError:
                 session.rollback()
                 raise
+        else:
+            # Make sure the new addition is flushed to, e.g., populate automatic primary keys.
+            session.flush()
 
         return self
 

--- a/aiida/storage/psql_dos/orm/utils.py
+++ b/aiida/storage/psql_dos/orm/utils.py
@@ -120,6 +120,8 @@ class ModelWrapper:
             self.session.add(self._model)
             if not self._in_transaction():
                 self.session.commit()
+            else:
+                self.session.flush()
         except IntegrityError as exception:
             self.session.rollback()
             raise exceptions.IntegrityError(str(exception))

--- a/tests/orm/nodes/test_node.py
+++ b/tests/orm/nodes/test_node.py
@@ -980,7 +980,7 @@ class TestNodeCaching:
         data.store()
 
         clone = data.clone()
-        clone._store_from_cache(data, with_transaction=True)  # pylint: disable=protected-access
+        clone._store_from_cache(data)  # pylint: disable=protected-access
 
         assert clone.is_stored
         assert clone.base.caching.get_cache_source() == data.uuid

--- a/tests/orm/test_querybuilder.py
+++ b/tests/orm/test_querybuilder.py
@@ -1495,9 +1495,30 @@ class TestConsistency:
             assert orm.load_node(pk).get_extra('key') == 'value'
 
     @pytest.mark.usefixtures('aiida_profile_clean')
-    @pytest.mark.skip('enable when https://github.com/aiidateam/aiida-core/issues/5802 is fixed')
     def test_iterall_with_store(self):
         """Test that nodes can be stored while being iterated using ``QueryBuilder.iterall``.
+
+        This is a regression test for https://github.com/aiidateam/aiida-core/issues/5802 .
+        """
+        count = 10
+        pks = []
+        pk_clones = []
+
+        for _ in range(count):
+            node = orm.Int().store()
+            pks.append(node.pk)
+
+        # Ensure that batch size is smaller than the total rows yielded
+        for [node] in orm.QueryBuilder().append(orm.Int).iterall(batch_size=2):
+            clone = orm.Int(node.value).store()
+            pk_clones.append(clone.pk)
+
+        for pk, pk_clone in zip(pks, sorted(pk_clones)):
+            assert orm.load_node(pk) == orm.load_node(pk_clone)
+
+    @pytest.mark.usefixtures('aiida_profile_clean')
+    def test_iterall_with_store_group(self):
+        """Test that nodes can be stored and added to groups while being iterated using ``QueryBuilder.iterall``.
 
         This is a regression test for https://github.com/aiidateam/aiida-core/issues/5802 .
         """
@@ -1510,7 +1531,7 @@ class TestConsistency:
             pks.append(node.pk)
 
         # Ensure that batch size is smaller than the total rows yielded
-        for [node] in orm.QueryBuilder().append(orm.Data).iterall(batch_size=2):
+        for [node] in orm.QueryBuilder().append(orm.Int).iterall(batch_size=2):
             clone = copy.deepcopy(node)
             clone.store()
             pks_clone.append((clone.value, clone.pk))

